### ISSUE TITLE
Conway spec proofreading

### ISSUE THREAD

**Issue by WhatisRT** [IntersectMBO/formal-ledger-specifications#457]
This is an issue for tracking proofreading of the Conway spec. We should look specifically for:

- Is every name/symbol defined/explained (within reason, possibly in older specs)?
- Is every name/symbol that's defined/explained also used?
- Do the figure names and groupings make sense?
- Any other irregularities (e.g. bad typesetting) in the figures?
- Are all features properly explained on a high level?
- Is the purpose of every check/restriction clear?
- Any typos/grammar issues?

Any violation of these should be documented here as a comment. It makes sense to fix small things immediately and have one PR that fixes them. Bigger things can become their own issues later.

## Issues found during proofread of Conway pdf

The items below are more concise versions of the items listed in the "Conway proofread" comment below.  (See the latter for more details.)

All line numbers below correspond to [commit 52e3113](https://github.com/IntersectMBO/formal-ledger-specifications/commit/52e31136cdfb6e38130cacaed75392fece71c0a2).

- [x] Figures
   - [x] `PParams.lagda` [Line 96](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L96): "Figure 3: Protocol parameter declarations"---fix `costmdls` line.
   - [x] `Utxo.lagda` [Lines 89--99](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo.lagda#L89C1-L100C1): first version of sentence, "Figures`~\ref{fig:supportfunctions:utxo}`,`~\ref{fig:functions:utxo}`, and`~\ref{fig:functions:utxo2}`..." best for Conway pdf
   - [x] `Utxo.lagda` [Lines 188--266](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo.lagda#L188C1-L266C14) fit "Figure 12: Functions..." on previous page
   - [x] `Gov.lagda` [Lines 93--96](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L93C1-L96C114): footnote on caption of Fig 15 out of place, move to Notation
   - [x] `Gov.lagda` [Lines 99--108](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L99C1-L108C79): hide definitions of
`getAidPairsList` and `_connects_to_` and explain them
   - [x] `Gov.lagda` [Lines 107--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L107C1-L113C97): fit in figure box
   - [x] `Certs.lagda` [Line 162](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Certs.lagda#L162): `\label{fig:sts:aux-cert}` to `\label{fig:sts:aux-cert-gov}`
   - [x] `Certs.lagda` [Lines 309--351](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Certs.lagda#L309C1-L351C14): Figure 24 "CERTS rules" too large
   - [x] `Ledger.lagda` [Lines 78--108](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ledger.lagda#L78C1-L108C14): Figure 27 "LEDGER transition system" bullet points misaligned
   - [x] `Enact.lagda` [Lines 106--154](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Enact.lagda#L106C1-L153C22): Figure 29 "ENACT..." too big
   - [x] `Ratify.lagda` [Lines 120--126](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ratify.lagda#L120C1-L126C65): explain table presenting `threshold`
   - [x] `Epoch.lagda` [Line 37](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Epoch.lagda#L37): `TODO` comment
   - [x] `Epoch.lagda` [Lines 69--80](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Epoch.lagda#L69C1-L81C1): (punctuation)
   - [x] `Epoch.lagda` [Lines 85--87](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Epoch.lagda#L85C3-L87C64): "Fig 37: EPOCH..." vertical vector alignment
   - [x] `Chain.lagda` [Line 105](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Chain.lagda#L105) of `Chain.lagda`: "Fig 40: CHAIN..." has line too long

- [x] Features and checks
   - [x] `Introduction.lagda` [Line 137](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L137): "...step relation" undefined
   - [x] `Transaction.lagda` [Line 49](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Transaction.lagda#L49): "...the hash of the serialized form..." to what does "the hash" refer?
   - [x] `Utxow.lagda` [Lines 18--19](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxow.lagda#L18C1-L19C12): "Figure 14 defines functions used for witnessing" too terse; what is meant by "witnessing."
   - [x] `Certs.lagda` [Lines 170--178](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Certs.lagda#L170C1-L178C72): meaning of "hot"/"cold"
   - [x] `Utxo/Properties.lagda` [Line 499](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo/Properties.lagda#L499): "...the UTxO relation is computable..." say more or ref Sec 1.4
   - [x] Appendix A seems scant
- [x] Typos and grammar
   - [x] Capitalization of section titles
   - [x] `Introduction.lagda` [Lines 64--65](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L64): "...impossible to replace" to "...impossible to avoid"
   - [x] `Introduction.lagda` [Lines 77--79](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L77): (punctuation) semicolon or period at end of lines
   - [x] `Introduction.lagda` [Line 133](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L133): "...the node..." to "...the nodes..."
   - [x] `Introduction.lagda` [Lines 165--166](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L165): "...necessarily right-unique..." to "...necessarily a (partial) function..."
   - [x] `Introduction.lagda` [Line 171](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L171): "...irrelevant - any..." to "...irrelevant---any..."
   - [x] `Introduction.lagda` [Line 190](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L190): "...sense - Agda..." to "...sense---Agda..."
   - [x] `Introduction.lagda` [Line 196](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L196): "...not defined everywhere..." to "...not necessarily defined..."
   - [x] `Introduction.lagda` [Lines 255--258](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L255C1-L258C59): relegate paragraph to footnote
   - [x] `Introduction.lagda` [Lines 262--263](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L262C9-L263C37): relegate "We also plan on refactoring..." to footnote.
   - [x] `Notation.lagda` [Line 9](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L9): "In this section, we introduce some notations used in this document." to "This section introduces some of the notation we use in this document and in our Agda formalization."
   - [x] `Notation.lagda` [Lines 36--37](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L36C66-L37C80): "Corestriction or range restriction is denoted the same..." to "Corestriction or range restriction is denoted similarly..."
   - [x] `GovernanceActions.lagda` [Lines 6, 10](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L6C2-L10C53): (punctuation) end 1 and 2 with semicolons; end item 3 with period
   - [x] `GovernanceActions.lagda` [Line 86](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L86) should end with a semicolon, not a period.
   - [x] `GovernanceActions.lagda` [Lines 112--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L112C125-L113C58): "...obsoleting nodes that are unable..." to "...rendering a node obsolete if it is unable..."
   - [x] `GovernanceActions.lagda` [Lines 125--127](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L125C1-L127C65): "For some types of...was submitted." to "For some governance actions, in addition to obtaining the necessary votes, enactment requires that the following condition is also satisfied: the state obtained by enacting the proposal is in fact the state that was intended when the proposal was submitted."
   - [x] `Utxo.lagda` [Lines 342--344](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo.lagda#L342C1-L344C25): smooth transition into paragraph; indicate we've finished talking about Fig 11
   - [x] `Gov.lagda` [Line 210](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L210): "`GovState` behaves similar to..." could be improved as either "`GovState` behaves like..."
   - [x] `Gov.lagda` [Line 222](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L222): "`validHFAction` is the property whether..." could be improved as "The `validHFAction` property indicates whether..."
   - [x] `Ratify.lagda` [Line 35](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ratify.lagda#L35): "...the (in principle arbitrary) semantic..." might be easier to read than "...the in principle arbitrary semantic..."
   - [x] `Utxo/Properties.lagda` [Lines 531--541](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo/Properties.lagda#L531C2-L541C65): vectorize constructors by adding `Utxo/Properties` to list of post-processed files in `Makefile`having problems processing this file.
   - [x] `EssentialAgda.lagda` [Line 1](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/EssentialAgda.lagda#L1): The section name "Appendix: Agda essentials" should be "Agda essentials".
   - [x] `PDF/ConwayBootstrap.lagda` [Lines 10--19](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PDF/ConwayBootstrap.lagda#L10C1-L19C14): punctuation

**Comment by williamdemeo** [comment]
# Conway proofread

N.B. This comment will be updated as the proofreading progresses.

All line numbers below correspond to [commit 52e3113](https://github.com/IntersectMBO/formal-ledger-specifications/commit/52e31136cdfb6e38130cacaed75392fece71c0a2).

--------------------------------------

## Names/symbols defined/explained

This section addresses the following concerns:

+ Is every name/symbol defined/explained (within reason, possibly in older specs)?
+ Is every name/symbol that's defined/explained also used?


### Section 1: Introduction

- [Lines 83--86](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L82C1-L87C1): The symbol `X` is not defined.
- [Line 137](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L137): "...over the step relation."  I believe this is the first time we call it a "step relation," though probably the reader can figure out we mean by this the state transition relation.
- [Line 200](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L200): We could add that the figure also defines equality on sets and the partial function type and the notation we use for these concepts.
- [Lines 198--219](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L198C16-L219C22): For some reason the powerset symbol is typeset with the `\AgdaBound{ℙ}` macro in the figure, while in the prose and in figures later in the document, it's set as `\AgdaFunction{ℙ}`.


### Section 3: Cryptographic primitives

- [Line 33](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Crypto.lagda#L33) The notation `Σ[_∈_]` is not defined.


### Section 5: Token algebras 

- [Line 28](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/TokenAlgebra.lagda#L28): `Type₁` is not defined.

- [Line 30](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/TokenAlgebra.lagda#L30): `⦃ Value-IsCommutativeMonoid' ⦄`  This is the first time
`⦃_⦄` is used.  It's not defined. (I'm not sure we want to define it; I'm just noting that it's not defined.)

- [Line 52](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/TokenAlgebra.lagda#L52): `_≗_` is not defined and the accompanying comment (`FIXME: rename!`) should be addressed and removed.


### Section 6: Addresses

- [Line 45](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Address.lagda#L45) `data Credential : Type where` This is the first use of `data` in the pdf; it's not defined.
- [Lines 53--63](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Address.lagda#L53C3-L63C25): The `field` keyword should be removed with the merge of [PR 462](https://github.com/IntersectMBO/formal-ledger-specifications/pull/462).
- [Lines 71--74](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Address.lagda#L71C1-L74C72): The notation `Σ[_∈_]` is not defined.

### Section 7: Scripts

- [Lines 88--91](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Script.lagda#L88C2-L91C52): `All` and `Any` are not defined.  (Should they be?  Maybe their meanings are obvious.)
- [Lines 92 ](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Script.lagda#L92) `MOf` was not defined.

### Section 8: Protocol parameters

- [Lines 33--34](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L33C1-L35C1):  `NewEpochState` is mentioned for the first time here, but its definition doesn't appear until much later, in Section 17 ("Epoch boundary"); add `(\ref{sec:17})` after the first mention of `NewEpochState`, and add the label `\label{sec:17}` to Sec 17.

### Section 9: Governance actions

- [Lines 142--144](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L142C53-L144C18): We explain that ⊤ is the unit type here, but this symbol appears earlier, in [Line 64](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L64) of `GovernanceActions.lagda` (Figure 11 "Governance actions"). 

-------------------

## Figures

This section addresses the following concerns:

+ Do the figure names and groupings make sense?
+ Any other irregularities (e.g. bad typesetting) in the figures?

### Section 6: Addresses

- [Lines 12--15](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Address.lagda#L12C1-L15C1) of `Address.lagda`: It's unfortunate we have just two sentences and then a large gap until Figure 6 appears on the next page.  Can we avoid this?

### Section 12: Governance

- [Line 107](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L107) and [Lines 110--114](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L110C1-L114C1) of `Gov.lagda`: Figure 28 "Enactability predicate" has four lines that escape the box.

### Section 13: Certificates

- [Lines 309--351](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Certs.lagda#L309C1-L351C14) of `Certs.lagda`: Figure 36 "CERTS rules" is too large to fit on the page without overlapping with the page number.  We should either split up the figure, reduce white space in the figure, or suppress the page number on that page. 

### Section 14: Ledger State Transition

- [Lines 78--108](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ledger.lagda#L78C1-L108C14) of `Ledger.lagda`: Figure 39 "LEDGER transition system" bullet points with vertical vectors are misaligned.

### Section 15: Enactment

- [Lines 106--154](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Enact.lagda#L106C1-L153C22) of `Enact.lagda`: Figure 42 "ENACT transition system" is too big to fit on a page because of the new vertical vectors; we should split this up or find a way to compress white space or simply don't use vertical vectors here.

### Section 18: Blockchain layer

- [Line 105](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Chain.lagda#L105) of `Chain.lagda`: line too long to fit in figure box. 

-------------------------

## Features and checks

This section addresses the following concerns:

+ Are all features properly explained on a high level?
+ Is the purpose of every check/restriction clear?


-------------------------

## Typos, grammar issues, and general comments

This section addresses the following concern:

+ Any typos/grammar issues?

### General comments

Capitalization of section titles is not consistent.  

+ Section 14: "Ledger State Transition" should be "Ledger state transition"
+ Appendix C: "Bootstrapping the Governance System" should be "Bootstrapping the governance system"


### Section 1: Introduction

- [Lines 64--65](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L64): "are very difficult or impossible to replace." --> "are very difficult or impossible to avoid."
- [Lines 77--79](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L77): I would add semicolons to the end of each item.
- [Lines 97--102](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L97): I'm not sure we should use the term "closure operation." Are we really taking a "closure?"  A closure operator is usually defined as an extensive, monotone, idempotent operator on sets.  What are the sets here?  Are they pairs of states?  How does the list of signals needed here fit into the usual definition of a closure operator?
- [Line 133](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L133): "...to be evaluated by the node..." should be "...to be evaluated by the nodes..."  (There isn't one node that we would call *the* node, is there?)
- [Lines 165--166](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L165): "...the step relation is necessarily right-unique..." should be "...the step relation is necessarily a function..." (much clearer and simpler)
- [Line 171](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L171): "...is irrelevant - any two..." should be either "...is irrelevant--any two..." or "...is irrelevant---any two..." to consistently follow standard hyphen conventions (of, e.g, academic journals)
- [Line 190](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L190): "...in a technical sense - Agda has..." should be "...in a technical sense---Agda has..."
- [Line 196](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L196): "i.e. functions not defined everywhere" should be "i.e., functions which are not necessarily defined everywhere"
- [Lines 248--249](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L248): This paragraph could be relegated to a footnote or removed; similarly for lines 658--659.


### Section 2: Notation

Should Section 2 ("Notation") be a subsection of the introduction so it can appear before we start using any of the notation described in that section?

[Line 9](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L9): I recommend changing "In this section, we introduce some notations used in this document." to "This section introduces some of the notation we use in this document and in our Agda formalization."

[Lines 36--37](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L36C66-L37C80): "Corestriction or range restriction is denoted the same..." should be "Corestriction or range restriction is denoted similarly..."

[Lines 38--39](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L38C1-L39C70): Do we really need to mention this?  It's very standard notation and
ought to be known by even those with a modest math background.

[Lines 42--43](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L42C71-L43C77): "while key-value pairs in \AgdaBound{m'} will be in the union if and only if the keys don't collide." Perhaps the following is clearer? "while a key-value pair in \AgdaBound{m'} will be in the union if and only if the key is not also a key in \AgdaBound{m}."

### Section 4: Base types

Should we add a few sentences here?

### Section 5: Token algebras

Should we add a few sentences here?

### Section 8: Protocol parameters

- [Lines 312--314](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L312C1-L315C1): "These new parameters are..." should be changed to "These new thresholds are..." for clarity.

### Section 9: Governance actions

- [Lines 6, 8](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L6C2-L8C65): add semicolons at the ends of items 1 and 2.

- [Lines 112--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L112C125-L113C58) (in footnote): "effectively obsoleting nodes that are unable to handle the upgrade" should be "effectively rendering nodes that are unable to handle the upgrade obsolete" 

- [Lines 229--230](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L229C63-L230C47): "For a proposal to be valid, the proposal must be set..." should be "For a proposal to be valid it must be set..."

**Comment by williamdemeo** [comment]
The list above seems to be growing too quickly.  I'll try to be less pedantic going forward, making fewer non-essential comments. I can also remove some of the non-essential comments from the above list, but lmkwyt.

**Comment by WhatisRT** [comment]
Most of these points make sense and should be fixed. I've left some comments below. The things that aren't listed below are the ones I agree with fully and had nothing else to say about.

> ### Section 3: Protocol Parameters
> * [Line 96](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L96) of `PParams.lagda`: "Figure 3: Protocol parameter declarations"---I don't know the meaning of the commented `costmdls` line.

This is what the definition _should_ be, but we cannot use it because then we can't get `DecEq` for `PParams`. This is something to think about.

> ### Section 7: Governance
> * [Lines 93--96](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L93C1-L96C114) in `Gov.lagda`: The footnote on the caption of Fig 15 seems out of place, considering how many other symbols are left undefined; I recommend we simply remove it.

We should rather put it in the `Notation` section, with a bunch of other list notation.

> * [Lines 99--108](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L99C1-L108C79): It's probably safe to hide the definitions of
>   `getAidPairsList` and `_connects_to_`; they are named so that the reader can get at least a rough idea of what they likely do without seeing the definition.
> * [Lines 107--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L107C1-L113C97): These lines do not fit in the figure box.

Yes, but we also need to explain it in the prose then.

> ### Section 11: Ratification
> * [Lines 120--126](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ratify.lagda#L120C1-L126C65) of `Ratify.lagda`: The presentation of `threshold` became more confusing and less intuitive with the introduction of the double vertical bars.

We just need to explain how to read the table.

> ### Section 12: Epoch boundary
> * [Line 37](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Epoch.lagda#L37) in `Epoch.lagda`: Do we want the `TODO` comment to be visible in the pdf?

None of this has changed with Conway, so we should just hide the entire figure.

> ### Section 1: Introduction
> * [Lines 97--102](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L97): I'm not sure we should use the term "closure operation." Are we really taking a "closure?"  A closure operator is usually defined as an extensive, monotone, idempotent operator on sets.  What are the sets here?  Are they pairs of states?  How does the list of signals needed here fit into the usual definition of a closure operator?

We've had a discussion with Orestis about this last year, and this is the name we came up with. "closure" means to take the smallest superrelation that satisfies these properties.

> ### Section 2: Notation
> * [Lines 38--39](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L38C1-L39C70): Do we really need to mention this?  It's very standard notation and
>   ought to be known by even those with a modest math background.

It's better to be thorough here.

> ### Section 8: Certificates
> * [Lines 170--178](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Certs.lagda#L170C1-L178C72) in `Certs.lagda`: Do we ever say what is meant by a "hot" credential? Probably hot/cold keys are defined in Shelley; should we remind the reader?

There's no technical meaning to this, and it's not defined anywhere. The cold credential is only used to register a hot credential, and then the hot credential is used for voting. The idea is that the access to the cold credential is kept in a really secure location, while the hot credential is kept somewhere with convenient access. If the hot credential gets compromised, it can be changed using the cold credential.

> ### Section 11: Ratification
> * [Line 52](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ratify.lagda#L52): "To pass the action, the yes votes need to be over the threshold x." might be clearer as "for an action to pass, the number of yes votes must exceed the threshold x."  Is it correct? i.e., the _number_ of yes votes must exceed the threshold?  Perhaps not since, I suppose, the raw number of votes is not what we actually use in the end, is it?  Maybe we should say more about this here.

It's specifically not the number of votes, it's the stake associated with the votes.

> ### Section 13: Blockchain layer
> Should we say anything about the figures in this section?

We're not changing anything in this section in Conway, so we can actually just remove it from this spec entirely.

> ### Section 14: Properties
> * [Line 499](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo/Properties.lagda#L499) in `Utxo/Properties.lagda`: "Here, we state the fact that the UTxO relation is computable." Should we say more here? Perhaps we should at least insert a ref to Section 1.4 "Computational."

We might as well just hide it for now since it doesn't contain anything that's relevant to Conway.

> ### Appendices
> Appendix A seems pretty scant. Perhaps we should supplement it so that it lives up to the promise of the paragraph introducing it, or else we could simply remove the section entirely.

I'd leave it for now, we'll see how much it's necessary later.

> * [Line 133](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L133): "...to be evaluated by the node..." should be "...to be evaluated by the nodes..."  (There isn't one node that we would call _the_ node, is there?)

There's the "Node" as in the software, and this sentence somewhat applies to both. But it's fine to just write "nodes".

> * [Lines 165--166](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L165): "...the step relation is necessarily right-unique..." should be "...the step relation is necessarily a function..." (simpler and clearer)

A function is would also be total, so it'd have to be a partial function.

> * [Lines 255--258](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L255C1-L258C59): This paragraph could be relegated to a footnote or removed; similarly for the sentence that begins, "We also plan on refactoring..." at [lines 262--263](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L262C9-L263C37).

I wrote this for the audit, but I think it's still pretty valid. We could move it to the `Notation` section.

> ### Section 2: Notation
> Should Section 2 ("Notation") be a subsection of the introduction so it can appear before we start using any of the notation it defines?

I'm not sure about this yet, we can leave it for later.

> ### Section 3: Protocol parameters
> * [Lines 312--314](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L312C1-L315C1) of `PParams.lagda`: "These new parameters are..." could be clearer if  changed to "These new thresholds are..."

But not all of those parameters are thresholds.

> * [Lines 112--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L112C125-L113C58) (in footnote): "...effectively obsoleting nodes that are unable to handle the upgrade." could be improved as follows: "...effectively rendering a node obsolete if it is unable to handle the upgrade."

The entire footnote could also just be removed.

> * [Lines 229--230](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L229C63-L230C47): "For a proposal to be valid, the proposal must be set..." could be said more clearly as follows: "For a proposal to be valid it must be set..."

We're talking about two things here, the proposal and the deposit. So the current sentence is wrong and it should be some variation of "For a proposal to be valid, the deposit must be set..."

> * [Line 269](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L269): "...and the role is allowed to vote." could be improved as "...and the voter with the given role is allowed to vote." (Conflating a voter with their role might be confusing.)

The right to vote is specifically associated with a role, not a voter.

**Comment by williamdemeo** [comment]
> > ### Section 3: Protocol Parameters
> > 
> > * [Line 96](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L96) of `PParams.lagda`: "Figure 3: Protocol parameter declarations"---I don't know the meaning of the commented `costmdls` line.
> 
> This is what the definition _should_ be, but we cannot use it because then we can't get `DecEq` for `PParams`. This is something to think about.

- [X] The commented `costmdls` line is now hidden.

---

> > ### Section 7: Governance
> > 
> > * [Lines 93--96](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L93C1-L96C114) in `Gov.lagda`: The footnote on the caption of Fig 15 seems out of place, considering how many other symbols are left undefined; I recommend we simply remove it.
> 
> We should rather put it in the `Notation` section, with a bunch of other list notation.

- [X] Moved footnote to `Notation` section in an item for list notation.

---

> > * [Lines 99--108](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L99C1-L108C79): It's probably safe to hide the definitions of
> >   `getAidPairsList` and `_connects_to_`; they are named so that the reader can get at least a rough idea of what they likely do without seeing the definition.
> > * [Lines 107--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L107C1-L113C97): These lines do not fit in the figure box.
> 
> Yes, but we also need to explain it in the prose then.

- [X] Definitions of `getAidPairsList` and `_connects_to_` now hidden and explained in the prose. 

---

> > ### Section 11: Ratification
> > 
> > * [Lines 120--126](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ratify.lagda#L120C1-L126C65) of `Ratify.lagda`: The presentation of `threshold` became more confusing and less intuitive with the introduction of the double vertical bars.
> 
> We just need to explain how to read the table.

- [ ] TODO: Re-read explanation of table and make sure it's clear or else improve it.

---

> > ### Section 12: Epoch boundary
> > 
> > * [Line 37](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Epoch.lagda#L37) in `Epoch.lagda`: Do we want the `TODO` comment to be visible in the pdf?
> 
> None of this has changed with Conway, so we should just hide the entire figure.

- [X] figure removed

---

> > ### Section 1: Introduction
> > 
> > * [Lines 97--102](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L97): I'm not sure we should use the term "closure operation." Are we really taking a "closure?"  A closure operator is usually defined as an extensive, monotone, idempotent operator on sets.  What are the sets here?  Are they pairs of states?  How does the list of signals needed here fit into the usual definition of a closure operator?
> 
> We've had a discussion with Orestis about this last year, and this is the name we came up with. "closure" means to take the smallest superrelation that satisfies these properties.

- [X] leave it as is

---

> > ### Section 2: Notation
> > 
> > * [Lines 38--39](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Notation.lagda#L38C1-L39C70): Do we really need to mention this?  It's very standard notation and
> >   ought to be known by even those with a modest math background.
> 
> It's better to be thorough here.

- [X] leave it as is

---

> > ### Section 8: Certificates
> > 
> > * [Lines 170--178](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Certs.lagda#L170C1-L178C72) in `Certs.lagda`: Do we ever say what is meant by a "hot" credential? Probably hot/cold keys are defined in Shelley; should we remind the reader?
> 
> There's no technical meaning to this, and it's not defined anywhere. The cold credential is only used to register a hot credential, and then the hot credential is used for voting. The idea is that the access to the cold credential is kept in a really secure location, while the hot credential is kept somewhere with convenient access. If the hot credential gets compromised, it can be changed using the cold credential.

- [X] "hot" and "cold" explained in a footnote

---

> > ### Section 11: Ratification
> > 
> > * [Line 52](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Ratify.lagda#L52): "To pass the action, the yes votes need to be over the threshold x." might be clearer as "for an action to pass, the number of yes votes must exceed the threshold x."  Is it correct? i.e., the _number_ of yes votes must exceed the threshold?  Perhaps not since, I suppose, the raw number of votes is not what we actually use in the end, is it?  Maybe we should say more about this here.
> 
> It's specifically not the number of votes, it's the stake associated with the votes.
> 

- [ ] TODO: change to "for an action to pass, the stake associated with the yes votes must exceed the threshold x." 

---

> > ### Section 13: Blockchain layer
> > Should we say anything about the figures in this section?
> 
> We're not changing anything in this section in Conway, so we can actually just remove it from this spec entirely.

- [ ] TODO: remove Section 13. 

---

> > ### Section 14: Properties
> > 
> > * [Line 499](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Utxo/Properties.lagda#L499) in `Utxo/Properties.lagda`: "Here, we state the fact that the UTxO relation is computable." Should we say more here? Perhaps we should at least insert a ref to Section 1.4 "Computational."
> 
> We might as well just hide it for now since it doesn't contain anything that's relevant to Conway.

- [ ] TODO: remove Section 14. 

---

> > ### Appendices
> > Appendix A seems pretty scant. Perhaps we should supplement it so that it lives up to the promise of the paragraph introducing it, or else we could simply remove the section entirely.
> 
> I'd leave it for now, we'll see how much it's necessary later.

- [X] leave as is

---

> > * [Line 133](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L133): "...to be evaluated by the node..." should be "...to be evaluated by the nodes..."  (There isn't one node that we would call _the_ node, is there?)
> 
> There's the "Node" as in the software, and this sentence somewhat applies to both. But it's fine to just write "nodes".
 
- [X] changed to "nodes"

---

> > * [Lines 165--166](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L165): "...the step relation is necessarily right-unique..." should be "...the step relation is necessarily a function..." (simpler and clearer)
> 
> A function is would also be total, so it'd have to be a partial function.
 
- [X] changed to "...the state transition relation is necessarily a (partial) function; i.e., there is at most one possible final state for each input data."  

---

> > * [Lines 255--258](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L255C1-L258C59): This paragraph could be relegated to a footnote or removed; similarly for the sentence that begins, "We also plan on refactoring..." at [lines 262--263](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Introduction.lagda#L262C9-L263C37).
> 
> I wrote this for the audit, but I think it's still pretty valid. We could move it to the `Notation` section.

- [X] moved comments on refactoring to a footnote

---

> > ### Section 2: Notation
> > Should Section 2 ("Notation") be a subsection of the introduction so it can appear before we start using any of the notation it defines?
> 
> I'm not sure about this yet, we can leave it for later.
 
- [X] leave as is

---

> > ### Section 3: Protocol parameters
> > 
> > * [Lines 312--314](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/PParams.lagda#L312C1-L315C1) of `PParams.lagda`: "These new parameters are..." could be clearer if  changed to "These new thresholds are..."
> 
> But not all of those parameters are thresholds.

- [X] leave as is

---

> > * [Lines 112--113](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L112C125-L113C58) (in footnote): "...effectively obsoleting nodes that are unable to handle the upgrade." could be improved as follows: "...effectively rendering a node obsolete if it is unable to handle the upgrade."
> 
> The entire footnote could also just be removed.

- [X] revised footnote as follows: "There are many varying definitions of the term ``hard fork'' in the blockchain industry. Hard forks typically refer to non-backwards compatible updates of a network. In Cardano, we attach a bit more meaning to the definition by calling any upgrade that would lead to \emph{more blocks} being validated a ``hard fork'' and force nodes to comply with the new protocol version, effectively rendering a node obsolete if it is unable to handle the upgrade."

---

> > * [Lines 229--230](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/GovernanceActions.lagda#L229C63-L230C47): "For a proposal to be valid, the proposal must be set..." could be said more clearly as follows: "For a proposal to be valid it must be set..."
> 
> We're talking about two things here, the proposal and the deposit. So the current sentence is wrong and it should be some variation of "For a proposal to be valid, the deposit must be set..."

- [X] revised as follows: "For a proposal to be valid, the deposit must be set to the current value of \govActionDeposit." 

---

> > * [Line 269](https://github.com/IntersectMBO/formal-ledger-specifications/blob/52e31136cdfb6e38130cacaed75392fece71c0a2/src/Ledger/Gov.lagda#L269): "...and the role is allowed to vote." could be improved as "...and the voter with the given role is allowed to vote." (Conflating a voter with their role might be confusing.)
> 
> The right to vote is specifically associated with a role, not a voter.

- [X] leave as is
